// tailwindcss-classnames/lib/cli/core/ClassesGroupTemplateGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassesGroupTemplateGenerator = void 0;
var utils_1 = require("../utils");
var ClassesGroupTemplateGenerator = /** @class */ (function () {
    function ClassesGroupTemplateGenerator(group, groupName, configPrefix) {
        var _this = this;
        this.generate = function () {
            return _this.generateMembersStatements().join('\n\n') + '\n\n' + _this.generateGroupStatement();
        };
        this.generateMembersStatements = function () {
            return _this.members.map(function (member) {
                return "export type T" + utils_1.capitalizeFirstLetter(member) + " = " + utils_1.generateTypes(_this.group[member], _this.configPrefix) + ";";
            });
        };
        this.generateGroupStatement = function () {
            var getMembersStatementsReferences = function () {
                return _this.members.map(function (member) { return 'T' + utils_1.capitalizeFirstLetter(member); }).join('\n  | ');
            };
            return ("export type T" + utils_1.capitalizeFirstLetter(_this.groupName) + " =" +
                '\n  | ' +
                getMembersStatementsReferences() +
                '\n');
        };
        this.group = group;
        this.groupName = groupName;
        this.configPrefix = configPrefix;
        this.members = Object.keys(this.group);
    }
    return ClassesGroupTemplateGenerator;
}());
exports.ClassesGroupTemplateGenerator = ClassesGroupTemplateGenerator;
