// tailwindcss-classnames/lib/cli/core/GeneratedFileWriter.js
"use strict";
/* eslint-disable @typescript-eslint/restrict-template-expressions */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratedFileWriter = void 0;
var tslib_1 = require("tslib");
var fs_1 = require("fs");
var colors_1 = tslib_1.__importDefault(require("colors"));
var baseTemplateString_1 = require("../lib/baseTemplateString");
var ClassesGenerator_1 = require("./ClassesGenerator");
var GeneratedFileWriter = /** @class */ (function () {
    function GeneratedFileWriter(options) {
        var _this = this;
        this.configFileData = '';
        this.write = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.missingCliOptionError().length > 0)
                            return [2 /*return*/, console.error(this.missingCliOptionError().red)];
                        return [4 /*yield*/, this.readTailwindConfigFile()];
                    case 1:
                        _a.sent();
                        fs_1.promises.writeFile("" + this.outputFilename, this.generateFileContent(), 'utf8').catch(function (error) {
                            console.error(colors_1.default.red(error));
                        });
                        return [2 /*return*/];
                }
            });
        }); };
        this.missingCliOptionError = function () {
            if (!_this.configFilename)
                return 'tailwindcss config file name or path is not provided';
            if (!_this.outputFilename)
                return 'Please provide a valid filename to add generated types to it';
            if (_this.customClassesTypeName && !_this.customClassesFilename)
                return 'Please provide the file containing custom classes';
            if (_this.customClassesFilename && !_this.customClassesTypeName)
                return 'Please provide the name of the exported custom type';
            return '';
        };
        this.readTailwindConfigFile = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var _a, err_1;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        _a = this;
                        return [4 /*yield*/, fs_1.promises.readFile("./" + this.configFilename, { encoding: 'utf-8' })];
                    case 1:
                        _a.configFileData = _b.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        err_1 = _b.sent();
                        console.error(("Error Reading: \"./" + this.configFilename + "\"").red);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        }); };
        this.generateFileContent = function () {
            return baseTemplateString_1.baseTemplateString
                .replace(/___ALL_CLASSES___/g, new ClassesGenerator_1.ClassesGenerator(eval(_this.configFileData.replace(/(['"])?plugins(['"])? *: *\[(.*|\n)*?],?/g, ''))).generate())
                .replace(/CUSTOM_FORMS_PLUGIN_TYPE/g, _this.configFileData.includes('@tailwindcss/custom-forms')
                ? '\n  | TCustomFormsPluginClasses'
                : '')
                .replace(/TYPOGRAPHY_PLUGIN_TYPE/g, _this.configFileData.includes('@tailwindcss/typography')
                ? '\n  | TTypographyPluginClasses'
                : '')
                .replace(/IMPORTED_T_CUSTOM_CLASSES/g, _this.isCustomClassesAdded ? '\n  | TCustomClassesFromExternalFile' : '')
                .replace(/T_CUSTOM_CLASSES_IMPORT_STATEMENT/g, _this.isCustomClassesAdded
                ? "import {" + _this.customClassesTypeName + " as TCustomClassesFromExternalFile} from './" + _this.customClassesFilename + "';"
                : '');
        };
        this.configFilename = options.configFilename;
        this.outputFilename = options.outputFilename;
        this.customClassesFilename = options.customClassesFilename;
        this.customClassesTypeName = options.customClassesTypeName;
        this.isCustomClassesAdded = !!(this.customClassesFilename && this.customClassesTypeName);
    }
    return GeneratedFileWriter;
}());
exports.GeneratedFileWriter = GeneratedFileWriter;
