// tailwindcss-classnames/es/cli/lib/default-classes/index.js
import { __assign } from "tslib";
import { Accessibility } from './Accessibility';
import { Backgrounds } from './Backgrounds';
import { Borders } from './Borders';
import { Effects } from './Effects';
import { FlexBox } from './Flexbox';
import { Grid } from './Grid';
import { Interactivity } from './Interactivity';
import { Layout } from './Layout';
import { Sizing } from './Sizing';
import { Spacing } from './Spacing';
import { SVG } from './SVG';
import { Tables } from './Tables';
import { Transforms } from './Transforms';
import { TransitionsAndAnimations } from './TransitionsAndAnimations';
import { Typography } from './Typography';
export var AllClassesFlat = __assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({}, Accessibility), Backgrounds), Borders), Effects), FlexBox), Grid), Interactivity), Layout), Sizing), Spacing), SVG), Tables), Transforms), TransitionsAndAnimations), Typography);
// TODO: only keep non-configurable classes [not generated from tailwind.config.js]
export var AllClasses = {
    Accessibility: Accessibility,
    Backgrounds: Backgrounds,
    Borders: Borders,
    Effects: Effects,
    FlexBox: FlexBox,
    Grid: Grid,
    Interactivity: Interactivity,
    Layout: Layout,
    Sizing: Sizing,
    Spacing: Spacing,
    SVG: SVG,
    Tables: Tables,
    Transforms: Transforms,
    TransitionsAndAnimations: TransitionsAndAnimations,
    Typography: Typography,
};
