// tailwindcss-classnames/es/cli/lib/default-classes/Layout.js
var boxSizing = ['box-border', 'box-content'];
var display = [
    'block',
    'inline-block',
    'inline',
    'flex',
    'inline-flex',
    'grid',
    'inline-grid',
    'table',
    'table-row',
    'table-cell',
    'contents',
    'hidden',
];
var container = ['container'];
var float = ['float-right', 'float-left', 'float-none', 'clearfix'];
var clear = ['clear-left', 'clear-right', 'clear-both', 'clear-none'];
var objectFit = [
    'object-contain',
    'object-cover',
    'object-fill',
    'object-none',
    'object-scale-down',
];
var objectPosition = [
    'object-bottom',
    'object-center',
    'object-left',
    'object-left-bottom',
    'object-left-top',
    'object-right',
    'object-right-bottom',
    'object-right-top',
    'object-top',
];
var overflow = [
    'overflow-auto',
    'overflow-hidden',
    'overflow-visible',
    'overflow-scroll',
    'overflow-x-auto',
    'overflow-y-auto',
    'overflow-x-hidden',
    'overflow-y-hidden',
    'overflow-x-visible',
    'overflow-y-visible',
    'overflow-x-scroll',
    'overflow-y-scroll',
    'scrolling-touch',
    'scrolling-auto',
];
var overscrollBehavior = [
    'overscroll-auto',
    'overscroll-contain',
    'overscroll-none',
    'overscroll-y-auto',
    'overscroll-y-contain',
    'overscroll-y-none',
    'overscroll-x-auto',
    'overscroll-x-contain',
    'overscroll-x-none',
];
var position = ['static', 'fixed', 'absolute', 'relative', 'sticky'];
var inset = [
    'inset-0',
    'inset-y-0',
    'inset-x-0',
    'top-0',
    'right-0',
    'bottom-0',
    'left-0',
    'inset-auto',
    'inset-y-auto',
    'inset-x-auto',
    'top-auto',
    'bottom-auto',
    'left-auto',
    'right-auto',
];
var visibility = ['visible', 'invisible'];
var zIndex = ['z-0', 'z-10', 'z-20', 'z-30', 'z-40', 'z-50', 'z-auto'];
export var Layout = {
    display: display,
    boxSizing: boxSizing,
    container: container,
    float: float,
    clear: clear,
    objectFit: objectFit,
    objectPosition: objectPosition,
    overflow: overflow,
    overscrollBehavior: overscrollBehavior,
    position: position,
    inset: inset,
    visibility: visibility,
    zIndex: zIndex,
};
