// tailwindcss-classnames/es/cli/lib/default-classes/TransitionsAndAnimations.js
var transitionProperty = [
    'transition-none',
    'transition-all',
    'transition',
    'transition-colors',
    'transition-opacity',
    'transition-shadow',
    'transition-transform',
];
var transitionDuration = [
    'duration-75',
    'duration-100',
    'duration-150',
    'duration-200',
    'duration-300',
    'duration-500',
    'duration-700',
    'duration-1000',
];
var transitionDelay = [
    'delay-75',
    'delay-100',
    'delay-150',
    'delay-200',
    'delay-300',
    'delay-500',
    'delay-700',
    'delay-1000',
];
var transitionTimingFunction = ['ease-linear', 'ease-in', 'ease-out', 'ease-in-out'];
var animation = [
    'animate-none',
    'animate-spin',
    'animate-ping',
    'animate-pulse',
    'animate-bounce',
];
export var TransitionsAndAnimations = {
    transitionProperty: transitionProperty,
    transitionDuration: transitionDuration,
    transitionTimingFunction: transitionTimingFunction,
    transitionDelay: transitionDelay,
    animation: animation,
};
