// tailwindcss-classnames/es/cli/core/ClassesGroupTemplateGenerator.js
import { capitalizeFirstLetter, generateTypes } from '../utils';
var ClassesGroupTemplateGenerator = /** @class */ (function () {
    function ClassesGroupTemplateGenerator(group, groupName, configPrefix) {
        var _this = this;
        this.generate = function () {
            return _this.generateMembersStatements().join('\n\n') + '\n\n' + _this.generateGroupStatement();
        };
        this.generateMembersStatements = function () {
            return _this.members.map(function (member) {
                return "export type T" + capitalizeFirstLetter(member) + " = " + generateTypes(_this.group[member], _this.configPrefix) + ";";
            });
        };
        this.generateGroupStatement = function () {
            var getMembersStatementsReferences = function () {
                return _this.members.map(function (member) { return 'T' + capitalizeFirstLetter(member); }).join('\n  | ');
            };
            return ("export type T" + capitalizeFirstLetter(_this.groupName) + " =" +
                '\n  | ' +
                getMembersStatementsReferences() +
                '\n');
        };
        this.group = group;
        this.groupName = groupName;
        this.configPrefix = configPrefix;
        this.members = Object.keys(this.group);
    }
    return ClassesGroupTemplateGenerator;
}());
export { ClassesGroupTemplateGenerator };
